### DISCUSSION TITLE
How to compile and run an external program in MOOSE?

### DISCUSSION THREAD

**Discussion by Ethan-xj** [idaholab/moose#28249] [Q&A General]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hello, everyone.
I want to know how can I compile and run an external program in MOOSE? The code is using fortran. Should I put the codes in the folder src and compile directly?

Ethan

**Comment by GiudGiud** [comment]
Hello

You could try to do that but it's not very tidy. 
We usually drop external codes in a contrib folder then modify the Makefile to compile it there 

You can see examples of this with Cardinal on the neams-th-coe GitHub 

There are also examples of compiling Fortran in the solid mechanics module for the UMAT

**Reply by Ethan-xj** [reply]
Thank you Giud. I'll check the Cardinal code.
